### PR TITLE
Expose locks info in REST API

### PR DESCRIPTION
**Pull request by hakonhall**

### CODE REVIEW

File: zkfacade/src/main/java/com/yahoo/vespa/curator/ThreadLockInfo.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+    public String getThreadName() { return thread.getName(); }
+    public String getLockPath() { return lockPath; }
+    public List<LockInfo> getLockInfos() { return List.copyOf(lockInfos); }
+
+    /** Mutable method (see class doc) */
+    void invokingAcquire(Duration timeout) {
+        invokeAcquireCount.incrementAndGet();
+        inCriticalRegionCount.incrementAndGet();
+        lockInfos.add(LockInfo.invokingAcquire(lock.getHoldCount(), timeout));
```
Comment (mpolden) [inline zkfacade/src/main/java/com/yahoo/vespa/curator/ThreadLockInfo.java:75]: Is this done to track the re-entry count?

File: zkfacade/src/main/java/com/yahoo/vespa/curator/LockInfo.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Optional;
+
+/**
+ * Information about a lock.
+ *
+ * <p>Should be mutated by a single thread.  Other threads may see an inconsistent state of this instance.</p>
+ */
+public class LockInfo {
```
Comment (mpolden) [inline zkfacade/src/main/java/com/yahoo/vespa/curator/LockInfo.java:13]: Maybe `LockMetrics` is a better name for this?

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/LocksResponse.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import com.yahoo.slime.Slime;
+import com.yahoo.vespa.curator.LockInfo;
+import com.yahoo.vespa.curator.ThreadLockInfo;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.time.Instant;
+import java.util.List;
+
+public class LocksResponse extends HttpResponse {
```
Comment (mpolden) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/LocksResponse.java:16]: Missing author/javadoc. Applies to some of the other classes as well.

File: zkfacade/src/main/java/com/yahoo/vespa/curator/Lock.java [hunk 4]
```diff
@@ -32,12 +32,20 @@ public Lock(String lockPath, Curator curator) {
...
+        ThreadLockInfo threadLockInfo = getThreadLockInfo();
+        threadLockInfo.invokingAcquire(timeout);
         try {
-            if ( ! mutex.acquire(timeout.toMillis(), TimeUnit.MILLISECONDS))
+            if ( ! mutex.acquire(timeout.toMillis(), TimeUnit.MILLISECONDS)) {
+                threadLockInfo.acquireTimedOut();
+
                 throw new UncheckedTimeoutException("Timed out after waiting " + timeout +
-                                                    " to acquire lock '" + lockPath + "'");
+                        " to acquire lock '" + lockPath + "'");
```
Comment (mpolden) [inline zkfacade/src/main/java/com/yahoo/vespa/curator/Lock.java:42]: Unintentional reformatting?

**Review by mpolden — COMMENTED** [review]
Good start!

File: zkfacade/src/main/java/com/yahoo/vespa/curator/ThreadLockInfo.java (hunk 1 above)
Comment (hakonhall) [reply zkfacade/src/main/java/com/yahoo/vespa/curator/ThreadLockInfo.java:75]: The getHoldCount() ? Yes - just to sanity-check it

File: zkfacade/src/main/java/com/yahoo/vespa/curator/LockInfo.java (hunk 2 above)
Comment (hakonhall) [reply zkfacade/src/main/java/com/yahoo/vespa/curator/LockInfo.java:13]: Hm, adding a LockCounters now with all the AtomicIntegers - hold on to this thought until you see that: Perhaps *that* should be called LockMetrics?  Although we are also adding metrics later, and those classes tend to be named *Metrics.

File: zkfacade/src/main/java/com/yahoo/vespa/curator/Lock.java (hunk 4 above)
Comment (hakonhall) [reply zkfacade/src/main/java/com/yahoo/vespa/curator/Lock.java:42]: Fixed

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/LocksResponse.java (hunk 3 above)
Comment (hakonhall) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/LocksResponse.java:16]: Fixed

**Comment by hakonhall** [comment]
Verified ed06bab worked ok on a live config server, will add unit tests.

**Comment by hakonhall** [comment]
@mpolden , this is now ready for proper review.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/NodesV2ApiHandler.java [hunk 5]
```diff
@@ -116,6 +116,7 @@ private HttpResponse handleGET(HttpRequest request) {
         if (pathS.startsWith("/nodes/v2/state/")) return new NodesResponse(ResponseType.nodesInStateList, request, orchestrator, nodeRepository);
         if (pathS.startsWith("/nodes/v2/acl/")) return new NodeAclResponse(request, nodeRepository);
         if (pathS.equals(    "/nodes/v2/command/")) return new ResourceResponse(request.getUri(), "restart", "reboot");
+        if (pathS.equals(    "/nodes/v2/locks")) return new LocksResponse();
```
Comment (mpolden) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/NodesV2ApiHandler.java:119]: Fine for now, but I think we should extract this into a top-level `/debug/` API if we intend to keep it.

File: zkfacade/src/test/java/com/yahoo/vespa/curator/stats/LockInfoSamplesTest.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
+package com.yahoo.vespa.curator.stats;// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Comment (mpolden) [inline zkfacade/src/test/java/com/yahoo/vespa/curator/stats/LockInfoSamplesTest.java:1]: Bug in test template? Copyright should be on first line.

File: zkfacade/src/test/java/com/yahoo/vespa/curator/stats/LockTest.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
+package com.yahoo.vespa.curator.stats;// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Comment (mpolden) [inline zkfacade/src/test/java/com/yahoo/vespa/curator/stats/LockTest.java:1]: Same as above.

File: zkfacade/src/test/java/com/yahoo/vespa/curator/stats/LockInfoSamplesTest.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class LockInfoSamplesTest {
```
Comment (mpolden) [inline zkfacade/src/test/java/com/yahoo/vespa/curator/stats/LockInfoSamplesTest.java:14]: Missing `@author`.

File: zkfacade/src/test/java/com/yahoo/vespa/curator/stats/LockTest.java [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertSame;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class LockTest {
```
Comment (mpolden) [inline zkfacade/src/test/java/com/yahoo/vespa/curator/stats/LockTest.java:22]: Same as above.

File: zkfacade/src/main/java/com/yahoo/vespa/curator/stats/LockInfo.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+/**
+ * Information about a lock.
+ *
+ * <p>Should be mutated by a single thread, except {@link #fillStackTrace()} which can be
+ * invoked by any threads.  Other threads may see an inconsistent state of this instance.</p>
+ *
+ * @author hakon
+ */
+public class LockInfo {
```
Comment (mpolden) [inline zkfacade/src/main/java/com/yahoo/vespa/curator/stats/LockInfo.java:16]: How about `LockAttempt`? `Info` is very generic.

File: zkfacade/src/main/java/com/yahoo/vespa/curator/stats/ThreadLockInfo.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+
+/**
+ * This class contains process-wide statistics and information related to acquiring and releasing
+ * {@link Lock}.  Instances of this class contain information tied to a specific thread and lock path.
+ *
+ * <p>Instances of this class are thread-safe as long as foreign threads (!= this.thread) avoid mutable methods.</p>
+ *
+ * @author hakon
+ */
+public class ThreadLockInfo {
```
Comment (mpolden) [inline zkfacade/src/main/java/com/yahoo/vespa/curator/stats/ThreadLockInfo.java:21]: `ThreadLockStats`? Matches the javadoc.

File: zkfacade/src/main/java/com/yahoo/vespa/curator/Lock.java [hunk 12]
```diff
@@ -20,54 +20,50 @@
...
+            throw new UncheckedTimeoutException("Timed out after waiting " + timeout +
+                    " to acquire lock '" + lockPath + "'");
         }
+        threadLockInfo.lockAcquired(lockPath);
     }
 
-    private void release() {
+    @Override
+    public void close() {
+        ThreadLockInfo.getCurrentThreadLockInfo().lockReleased(lockPath);
```
Comment (mpolden) [inline zkfacade/src/main/java/com/yahoo/vespa/curator/Lock.java:59]: Shouldn't this happen after the `mutex.release()` below?

File: zkfacade/src/test/java/com/yahoo/vespa/curator/stats/LockInfoSamplesTest.java (hunk 6 above)
Comment (hakonhall) [reply zkfacade/src/test/java/com/yahoo/vespa/curator/stats/LockInfoSamplesTest.java:1]: Bug in IntelliJ/have never been able to figure it out.  Fixed.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/NodesV2ApiHandler.java (hunk 5 above)
Comment (hakonhall) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/NodesV2ApiHandler.java:119]: +1

File: zkfacade/src/test/java/com/yahoo/vespa/curator/stats/LockTest.java (hunk 7 above)
Comment (hakonhall) [reply zkfacade/src/test/java/com/yahoo/vespa/curator/stats/LockTest.java:1]: Fixed

File: zkfacade/src/test/java/com/yahoo/vespa/curator/stats/LockInfoSamplesTest.java (hunk 8 above)
Comment (hakonhall) [reply zkfacade/src/test/java/com/yahoo/vespa/curator/stats/LockInfoSamplesTest.java:14]: Fixed

File: zkfacade/src/test/java/com/yahoo/vespa/curator/stats/LockTest.java (hunk 9 above)
Comment (hakonhall) [reply zkfacade/src/test/java/com/yahoo/vespa/curator/stats/LockTest.java:22]: Fixed

File: zkfacade/src/main/java/com/yahoo/vespa/curator/stats/ThreadLockInfo.java (hunk 11 above)
Comment (hakonhall) [reply zkfacade/src/main/java/com/yahoo/vespa/curator/stats/ThreadLockInfo.java:21]: Done

File: zkfacade/src/main/java/com/yahoo/vespa/curator/stats/LockInfo.java (hunk 10 above)
Comment (hakonhall) [reply zkfacade/src/main/java/com/yahoo/vespa/curator/stats/LockInfo.java:16]: Done

File: zkfacade/src/main/java/com/yahoo/vespa/curator/Lock.java (hunk 12 above)
Comment (hakonhall) [reply zkfacade/src/main/java/com/yahoo/vespa/curator/Lock.java:59]: To get any Instant.now() to include the time in release?

Comment (mpolden) [reply zkfacade/src/main/java/com/yahoo/vespa/curator/Lock.java]: No, currently if `mutex.release()` fails  you will still have tagged it as released in `ThreadLockInfo`.

Comment (hakonhall) [reply zkfacade/src/main/java/com/yahoo/vespa/curator/Lock.java]: I could count an error, but the lock must be assumed to be released whether an exception is thrown or not

Comment (hakonhall) [reply zkfacade/src/main/java/com/yahoo/vespa/curator/Lock.java]: Added count

**Review by mpolden — APPROVED** [review]
👍

File: zkfacade/src/main/java/com/yahoo/vespa/curator/Lock.java [hunk 13]
```diff
@@ -21,51 +20,38 @@
  */
 public class Lock implements Mutex {
 
-    private final ReentrantLock lock;
     private final InterProcessLock mutex;
     private final String lockPath;
 
     public Lock(String lockPath, Curator curator) {
         this.lockPath = lockPath;
-        this.lock = new ReentrantLock(true);
```
[the remark this answers is not in the document]
Comment (jonmv) [reply zkfacade/src/main/java/com/yahoo/vespa/curator/Lock.java:29]: It _did_ trigger a few times, but this really only verified something we already suspected — that the same Curator lock may be held by several parties simultaneously as a result of connection issues, etc..

[the remark this answers is not in the document]
Comment (jonmv) [reply zkfacade/src/main/java/com/yahoo/vespa/curator/Lock.java:29]: (It also would prevent this intra-JVM, of course, but not inter-JVM.)

[the remark this answers is not in the document]
Comment (hakonhall) [reply zkfacade/src/main/java/com/yahoo/vespa/curator/Lock.java:29]: AFAIK `lock.tryLock()` can only return false after a successful `InterProcessMutex.acquire()` if there is a ZK connection loss between two `acquire()`, and the respective threads would then both appear to have the lock.  I didn't know that was observed.  But yeah, this guard would only guard intra-JVM.

**Pull request merged**